pubsub/gossipsub/gossipsub-v1.3.md

```diff
@@ -0,0 +1,166 @@
+
+See the [lifecycle document][lifecycle-spec] for context about maturity level and spec status.
+
+[lifecycle-spec]: https://github.com/libp2p/specs/blob/master/00-framework-01-spec-lifecycle.md
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+- [Overview](#overview)
+- [Motivation](#motivation)
+- [Parameters](#parameters)
+- [Observing a topic](#observing-a-topic)
+- [Notifying observing peers](#notifying-observing-peers)
+- [Unobserving a topic](#unobserving-a-topic)
+- [Limits on observing peers](#limits-on-observing-peers)
+- [Control Messages](#control-messages)
+  - [OBSERVE](#observe)
+  - [UNOBSERVE](#unobserve)
+- [Protobuf](#protobuf)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Overview
+
+This document specifies a minimal extension to the [gossipsub v1.2](gossipsub-v1.2.md) protocol.
+
+The proposed extension is intended to enable the nodes to get notified when there is a new message in
+a topic without actually receiving the actual message.
+
+Four new control messages are introduced: `OBSERVE`, `OBSERVED`, `UNOBSERVE`, and `UNOBSERVED`. They
+are primarily used to notify the peers that the node wants to observe/unobserve if there is a new
+message in the topic.
+
+## Motivation
+
+There are many use cases of topic observation:
+
+- For a node that just wants to get notified when there is a new message in the topic and doesn't
+  want to consume too much bandwidth by directly subscribing to the topic.
+
+- For a node that wants to subscribe to the topic, but has some bandwidth constraint. The node can
+  observe the topic and see which peer notifies it first and later send the `IWANT` to get the
+  actual message. This ensures that the bandwidth used is approximately the size of the message.
+
+The downside of topic observation is that the observing node is not part of the mesh network, so it
+only consumes the messages, but doesn't contribute to other peers by forwarding the messages.
+
+## Parameters
+
+This section lists new configurable parameters.
+
+| Parameter   | Description                                                | Reasonable Default |
+|-------------|------------------------------------------------------------|--------------------|
+| `D_observe` | (Optional) the maximum number of observing peers per topic | `D`                |
+
+## Observing a topic
+
+A node can observe a topic by sending an `OBSERVE` message to a peer that already subscribes to the topic.
+If the observation is successful, the peer will send back an `OBSERVED` message.
+
+If the node already subscribes to the topic, it doesn't make sense for that node to observe the topic so
+the node SHOULD NOT send an `OBSERVE` message for that topic.
+
+## Notifying observing peers
+
+When a node receives a message in the topic, it will fordward the message to its mesh peers as usual and
```

Comment:
In the context of sampling for peerDAS, the flow of data becomes trickier. 
-> Node receives its columns via custody subnets.
-> In the following slot(or any N slots into the future) it then decides to sample random peers for any number of columns.
-> It sends our the `OBSERVE` control message to peers subscribed to the desired column gossip topics. And the peer then sends back `OBSERVED` 

However here is the problem, the peer can only send the control message *after* the slot has passed or at least after it has received its custody columns. It means that the subcribed peer will not be tracking these particular messages when we actually need them to. So it wont be able to send the `IHAVE` messages that we want.

The way to get around this would be to have us sending `OBSERVE` messages for all column subnets all the time. It would mean that every node would constantly have to deal with data that it technically doesn't care about at all until it might need it in the future.

Reply:
> However here is the problem, the peer can only send the control message _after_ the slot has passed or at least after it has received its custody columns

Why? what I would like to do is to observe the subnets a moment before the slot and you will be notified during the slot. So the flow should be.

-> A moment before the slot, nodes decide the columns they want to sample.
-> Nodes send `OBSERVE` and receive `OBSERVED` to/from their peers in the corresponding subnets.
-> In the slot, nodes receives their custody columns
-> Observing nodes get notified

Reply:
I think joining a mesh for such a short period of time can be pretty unstable. One issue you will run into is that the peers that you want to `Observe` might not have the slots that you want. So you would need to constantly send out to all the available peers that you want to 'observe' and by the time you find a peer with a free slot it might be too late . To look at analogue for this, look at what happens when a node first subscribes to any topic. There will be a lot of `GRAFT` messages and `PRUNE` messages sent back, before they settle on a peer set that has the appropriate space in their mesh.

Reply:
> I think joining a mesh for such a short period of time can be pretty unstable

It's complete stable because observation is not subscription. You don't actually join the mesh and you are not actually in the middle of the network, but kind of edge of the network (because you don't forward messages to other peers).

> One issue you will run into is that the peers that you want to `Observe` might not have the slots that you want. So you would need to constantly send out to all the available peers that you want to 'observe' and by the time you find a peer with a free slot it might be too late . To look at analogue for this, look at what happens when a node first subscribes to any topic. There will be a lot of `GRAFT` messages and `PRUNE` messages sent back, before they settle on a peer set that has the appropriate space in their mesh.

Unlike subscription, observation is cheap (because it's only about sending IHAVEs). Together with https://github.com/libp2p/specs/pull/617#issuecomment-2139042308, I think we can set `D_observe` to a very high value, or remove it at all. Then, you will very likely have a free slot very fast or do not worry about it at all, if we remove it.

Reply:
Yeah but each peer has a limit for how many 'observers' they would allow for each topic. You run into the possibility, the peer has used up all of its slots for 'observation' and there is no more space for your node to observe message ids coming into the remote peer's mesh. Unless the plan is to allow many observers per topic where this stops being an issue.